src/encoding.ts

```diff
@@ -0,0 +1,94 @@
+/**
+ * encoding.js
+ *
+ * This NPM module provide encoding and decoding utilities for interacting with
+ * dFusion smart contracts where manual byte encoding was needed because of
+ * solidity ABI limitations.
+ */
+import BN from "bn.js";
+import assert from "assert";
+
+export interface Order<T = string> {
+  user: string;
+  sellTokenBalance: T;
+  buyToken: number;
+  sellToken: number;
+  validFrom: number;
+  validUntil: number;
+  priceNumerator: T;
+  priceDenominator: T;
+  remainingAmount: T;
+}
+
+export interface IndexedOrder<T> extends Order<T> {
+  orderId: number;
+}
+
+class OrderBuffer {
+  index = 2; // skip '0x'
+  constructor(public bytes: string) {}
```

Comment:
nit: Is there anything preventing this:
```suggestion
  constructor(private readonly bytes: string) {}
```

Reply:
**edit**: I see we are using the `buffer.bytes.length`, maybe this can be a `hasMoreBytes()` so the `index` and `bytes` implementation details can be hidden?

Reply:
Will do 👍